In [1]:
import pandas as pd
df = pd.read_csv("messy_sales.csv")

In [2]:
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [3]:
df.shape

(300, 6)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   order_id  300 non-null    int64  
 1   date      300 non-null    object 
 2   product   300 non-null    object 
 3   price     288 non-null    float64
 4   qty       300 non-null    int64  
 5   zip       300 non-null    int64  
dtypes: float64(1), int64(3), object(2)
memory usage: 14.2+ KB


In [5]:
df.describe()

,order_id,price,qty,zip
count,300.000000,288.000000,300.000000,300.000000
mean,1144.816667,56.068438,2.913333,45385.276667
std,84.515319,72.307182,1.560107,39345.608641
min,1000.000000,2.930000,-5.000000,2116.000000
25%,1071.750000,10.945000,2.000000,2134.000000
50%,1144.500000,37.530000,3.000000,30303.000000
75%,1217.250000,61.417500,4.000000,90210.000000
max,1291.000000,479.290000,5.000000,98101.000000


### Intial Obesrvations
Three things that I  see wrong  with this data  set  is: 
1.  The  price count is 288 but there are 300 rows, so there may be missing values.
2. The minimum for quantity is -5 which seems suspicious for a sales order.
3. The date column has inconsistent formats, some follow MM/DD/YYYY and others YYYY-MM-DD.

In [6]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [7]:
fill_value = df["price"].median()
df["price"] = df["price"].fillna(fill_value)

In [8]:
print (fill_value)

37.53


In [9]:
df.isna().sum()

order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [10]:
print(df.duplicated().sum())

8


In [11]:
df. shape

(300, 6)

In [12]:
df = df.drop_duplicates()

In [13]:
df. shape

(292, 6)

In [14]:
df["zip"] = df["zip"].astype(str).str.zfill(5)

In [15]:
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,37.53,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,37.53,4,30303
6,1204,04/18/2026,charger,8.41,2,02134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 292 entries, 0 to 299
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   order_id  292 non-null    int64  
 1   date      292 non-null    object 
 2   product   292 non-null    object 
 3   price     292 non-null    float64
 4   qty       292 non-null    int64  
 5   zip       292 non-null    object 
dtypes: float64(1), int64(2), object(3)
memory usage: 16.0+ KB


In [17]:
df["date"] = pd.to_datetime(df["date"], format="mixed")

In [18]:
df.dtypes

order_id             int64
date        datetime64[ns]
product             object
price              float64
qty                  int64
zip                 object
dtype: object

## Checkpoint

In [19]:
df. shape

(292, 6)

In [20]:
df. head(10)

,order_id,date,product,price,qty,zip
0,1254,2026-04-06,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,2026-05-06,mug,5.50,1,10001
3,1066,2026-05-30,notebook,37.53,1,98101
4,1114,2026-04-06,webcam,45.59,4,90405
5,1280,2026-04-22,pen set,37.53,4,30303
6,1204,2026-04-18,charger,8.41,2,02134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [21]:
df.dtypes

order_id             int64
date        datetime64[ns]
product             object
price              float64
qty                  int64
zip                 object
dtype: object

In [22]:
df[df["qty"] < 0]

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


In [23]:
df. to_csv ("sales_clean.csv", index=False)

## Cleaning Log

- There was a total of 300 rows loaded
- Filled 12 missing price values with median 37.53.
- The maximum price is 479.29, the mean is 56.07 and median is 37.53.
  I chose the median because the high price can skew the data and mean.
- Removed 8 duplicate rows leaving 292.
- Converted zip codes from integers to string and restored leading zeroes by adapting to 5-character zips.
- Converted the date column from object to real datetimes and standardized the mixed formats.
- Decided to  treat the 3 negative quantities (-5,-4,-2) as possiblle returns instead of typos. I decided to keep these the rows instead of correcting or deleting, since return of items is possiblele.
- A wrong choice at any of these steps could mislead a business decision becuase it could give a false idea of revenue or even the amount of a product that is sold. For example, if i decided to keep the duplicated rows that would increase the total revenue and product quantity.

Git Hub Link: https://github.com/jazzminowens/cs82a-portfolio